### PR TITLE
Allow Javascript expressions as conditions for edges

### PR DESCRIPTION
**Pull request by jyeshe**
## Notes for the reviewer

Adds the option for JS expression on condition, validates it as an expression and save to be passed to the worker.

Screenshot from 2023-12-07 18-31-38

Note: E2E test gonna be implemented on a separate PR.

## Related issue

Refs #1498

## Review checklist

- [x] I have performed a **self-review** of my code
- [x] I have verified that all appropriate **authorization policies** have been implemented and tested
- [x] If needed, I have updated the **changelog**
- [ ] Product has **QA'd** this feature

### CODE REVIEW

**Review by taylordowns2000 — CHANGES_REQUESTED** [review]
hey @jyeshe , i'm seeing at least two things here that aren't quite right:
1. i'm not able to modify the edge between a trigger and a job. the options there should be always and the new custom js option.
2. after adding my custom condition, it disappears. (both the label and the input box.) and i'm unable to access it again.

https://www.loom.com/share/edd41b08c7a745849e9dff56e3dc8ba9

finally, please rebase this on `main` as i think it's a bit out of date and a rebase will make testing easier

**Comment by jyeshe** [comment]
> hey @jyeshe , i'm seeing at least two things here that aren't quite right:
> 
> 1. i'm not able to modify the edge between a trigger and a job. the options there should be always and the new custom js option.
> 2. after adding my custom condition, it disappears. (both the label and the input box.) and i'm unable to access it again.
> 
> https://www.loom.com/share/edd41b08c7a745849e9dff56e3dc8ba9
> 
> finally, please rebase this on `main` as i think it's a bit out of date and a rebase will make testing easier

Morning @taylordowns2000! I considered the first item as out of the scope since it mingles with another app feature, whether the job is executed or not when the trigger is enabled but as there was no exception mentioned before it makes sense to apply  the additional condition to this edge.
For the second item, sorry for the inconvenience, I have relied to much on the automated test case. It works now (labels are preserved during and after edit/save).

For the rebase there was a very recent conflict resolution that brought main changes (8b761f2) but it is also rebased now.

**Comment by stuartc** [comment]
@jyeshe this is great! I added the two small recommendations in the comments - I also made some some small component/layout changes; I reverted the `font-medium` stuff as well... The app isn't consistent at the moment in this area so I'd rather have it look less than perfect - and change _all_ the components at once than introduce a new variant.

File: lib/lightning/workflows/edge.ex [hunk 1]
```diff
@@ -80,20 +85,52 @@ defmodule Lightning.Workflows.Edge do
...
-    |> Enum.map(&get_field(changeset, &1))
-    |> case do
-      [trigger, _condition] when not is_nil(trigger) ->
-        changeset
-        |> validate_inclusion(:condition, [:always],
-          message: "must be :always when source is a trigger"
+    if nil != get_field(changeset, :source_trigger_id) do
+      changeset
+      |> validate_inclusion(:condition, [:always, :js_expression],
+        message: "must be :always or :js_expression when source is a trigger"
```
[the remark this answers is not in the document]
Comment (jyeshe) [reply lib/lightning/workflows/edge.ex:96]: Better for user indeed

**Comment by jyeshe** [comment]
> @jyeshe this is great! I added the two small recommendations in the comments - I also made some some small component/layout changes; I reverted the `font-medium` stuff as well... The app isn't consistent at the moment in this area so I'd rather have it look less than perfect - and change _all_ the components at once than introduce a new variant.

I have customized for the figma style but it makes sense to change for all screens at once.

**Comment by stuartc** [comment]
> I have customized for the figma style but it makes sense to change for all screens at once.

Awesome thanks, the Figma styles are "idealistic", there's been no endeavour to standardise any of this. 

I mean we do try, but yeah we don't make any effort to either maintain or change if you know what I mean.

So case by case right now.

**Comment by stuartc** [comment]
@jyeshe I was expecting a validation on the model as well, bonus points for the inputs though, but we should have on the model in case params slip through past the inputs, or they are ever sent via an API etc.

**Comment by jyeshe** [comment]
> @jyeshe I was expecting a validation on the model as well, bonus points for the inputs though, but we should have on the model in case params slip through past the inputs, or they are ever sent via an API etc.

It was in a separate commit c5cf7b8

**Comment by stuartc** [comment]
@jyeshe can you rebase and squash/tidy up the commits and I'll merge it 👍

**Review by stuartc — APPROVED** [review]

**Comment by stuartc** [comment]
@jyeshe I just merged the rename brach, if you could rebase and keep that rename stuff in a single and separate commit if you'd like.

**Pull request closed without merging**